### PR TITLE
Display energy & num selected in HUD

### PR DESCRIPTION
**Pull request by JackCrumpLeys**
Implements part of #397. Adds on to #483 and works towards #473.

### CODE REVIEW

File: crates/controller/src/hud/details.rs [hunk 1]
```diff
@@ -1,37 +1,113 @@
...
+            battery_total += battery.energy().floor() as i64;
+            battery_max += battery.capacity().floor() as i64;
+        }
+    }
+
+    if battery_max == 0 {
+        return;
+    }
+
+    let text = format!(
```
Comment (Indy2222) [inline crates/controller/src/hud/details.rs:100]: With:

```rust
const PREFIXES: [&'static str; 5] = ["T", "G", "M", "k", ""];
```

and

```rust
fn format_units(value: f64, units: &str) -> String {
    for i in (0..PREFIXES.len()).rev() {
        let coeff = 1000f64.powi(i as i32);
        if value > coeff {
            let rounded = most_significant(value / coeff);
            return format!("{}{}{units}", rounded, PREFIXES[i]);
        }
        
    }
    
    unreachable!("TODO");
}

fn most_significant(value: f64) -> f64 {
    if value == 0.0 {
        0.0
    } else {
        let d = 3 - value.abs().log10().ceil() as i32;
        let p = 10f64.powi(d);
        (value * p).round() / p
    }
}
```

we can use pretty `format_units(battery_max, "J")` and `format_units(battery_total, "J")`.

File: crates/controller/src/hud/details.rs [hunk 2]
```diff
@@ -1,37 +1,113 @@
...
+        battery_total,
+        battery_max,
+        battery_total * 100 / battery_max,
+        selected.iter().count()
+    );
+
+    text_ops
+        .set_text(ui.0, text)
+        .expect("Failed to set text of details");
+    #[derive(Resource)]
```
Comment (Indy2222) [inline crates/controller/src/hud/details.rs:111]: What's this?

**Review by Indy2222 — CHANGES_REQUESTED** [review]

File: crates/gui/src/body_text.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                ..default()
+            },
+            BodyText,
+        ));
+
+        commands.with_children(|builder| {
+            builder.spawn(TextBundle::from_section(caption, text_style));
+        });
+
+        commands
```
[the remark this answers is not in the document]
Comment (JackCrumpLeys) [reply crates/gui/src/body_text.rs:46]: it follows the same design as the Button component. It lets you string multiple commands together.

File: crates/controller/src/hud/details.rs (hunk 1 above)
Comment (JackCrumpLeys) [reply crates/controller/src/hud/details.rs:100]: Very pretty!

File: crates/controller/src/hud/details.rs [hunk 4]
```diff
@@ -1,37 +1,130 @@
...
+use crate::selection::Selected;
+
+const PREFIXES: [&str; 5] = ["T", "G", "M", "k", ""];
 
 pub(crate) struct DetailsPlugin;
 
 impl Plugin for DetailsPlugin {
     fn build(&self, app: &mut App) {
-        app.add_system(spawn_details.in_schedule(OnEnter(GameState::Playing)));
+        app.add_system(setup.in_schedule(OnEnter(GameState::Playing)))
```
Comment (Indy2222) [inline crates/controller/src/hud/details.rs:16]: Please add a cleanup system which removes `DetailsText` after the game is finished.

File: crates/controller/src/hud/details.rs [hunk 5]
```diff
@@ -1,37 +1,130 @@
...
+
+    if battery_max == 0. {
+        text_ops
+            .set_text(ui.0, "")
+            .expect("Failed to set text of details");
+        return;
+    }
+
+    let text = format!(
+        "Battery: Battery: {} / {} ({:.1}%)\nSelected {}",
```
Comment (Indy2222) [inline crates/controller/src/hud/details.rs:120]: `Battery: Battery:` -> `Battery:` (IMO a typo).

File: crates/controller/src/hud/details.rs [hunk 6]
```diff
@@ -1,37 +1,130 @@
...
-    ));
+            "",
+        )
+        .id();
+    commands.entity(node).add_child(details_text);
+
+    commands.insert_resource(DetailsText(details_text));
+}
+
+fn format_units(value: f64, units: &str) -> String {
```
Comment (Indy2222) [inline crates/controller/src/hud/details.rs:70]: I would write unnittests for this one: with at least 0, smaller than 0 and a few larger numbers.

**Review by Indy2222 — CHANGES_REQUESTED** [review]

**Comment by Indy2222** [comment]
The font is way too large on my computer. Can you make it smaller?

image

**Comment by JackCrumpLeys** [comment]
font size 30->16

**Review by Indy2222 — APPROVED** [review]

**Pull request merged**